taskchampion/src/storage/sqlite.rs

```diff
@@ -1,355 +1,346 @@
-        &self.storage.tasks_bucket
+/// Stores [`Operation`] in SQLite
+impl FromSql for Operation {
+    fn column_result(value: rusqlite::types::ValueRef<'_>) -> rusqlite::types::FromSqlResult<Self> {
+        let o: Operation = serde_json::from_str(value.as_str()?)
+            .map_err(|_| rusqlite::types::FromSqlError::InvalidType)?;
+        Ok(o)
     }
-    fn numbers_bucket(&self) -> &'t Bucket<'t, Integer, ValueBuf<Msgpack<u64>>> {
-        &self.storage.numbers_bucket
+}
+
+/// Parsers Operation stored as JSON in string column
+impl ToSql for Operation {
+    fn to_sql(&self) -> rusqlite::Result<rusqlite::types::ToSqlOutput<'_>> {
+        let s = serde_json::to_string(&self)
+            .map_err(|e| rusqlite::Error::ToSqlConversionFailure(Box::new(e)))?;
+        Ok(s.into())
     }
-    fn uuids_bucket(&self) -> &'t Bucket<'t, Integer, ValueBuf<Msgpack<Uuid>>> {
-        &self.storage.uuids_bucket
+}
+
+/// SqliteStorage is an on-disk storage backed by SQLite3.
+pub struct SqliteStorage {
+    con: Connection,
+}
+
+impl SqliteStorage {
+    pub fn new<P: AsRef<Path>>(directory: P) -> anyhow::Result<SqliteStorage> {
+        // Ensure parent folder exists
+        std::fs::create_dir_all(&directory)?;
+
+        // Open (or create) database
+        let db_file = directory.as_ref().join("taskchampion.sqlite3");
+        let con = Connection::open(db_file)?;
+
+        // Initialize database
+        let queries = vec![
+            "CREATE TABLE IF NOT EXISTS operations (id INTEGER PRIMARY KEY AUTOINCREMENT, data STRING);",
+            "CREATE TABLE IF NOT EXISTS sync_meta (key STRING PRIMARY KEY, value STRING);",
+            "CREATE TABLE IF NOT EXISTS tasks (uuid STRING PRIMARY KEY, data STRING);",
+            "CREATE TABLE IF NOT EXISTS working_set (id INTEGER PRIMARY KEY, uuid STRING);",
+        ];
+        for q in queries {
+            con.execute(q, []).context("Creating table")?;
+        }
+
+        Ok(SqliteStorage { con })
     }
-    fn operations_bucket(&self) -> &'t Bucket<'t, Integer, ValueBuf<Msgpack<Operation>>> {
-        &self.storage.operations_bucket
+}
+
+struct Txn<'t> {
+    txn: Option<rusqlite::Transaction<'t>>,
+}
+
+impl<'t> Txn<'t> {
+    fn get_txn(&self) -> Result<&rusqlite::Transaction<'t>, SqliteError> {
+        self.txn
+            .as_ref()
+            .ok_or(SqliteError::TransactionAlreadyCommitted)
     }
-    fn working_set_bucket(&self) -> &'t Bucket<'t, Integer, ValueBuf<Msgpack<Uuid>>> {
-        &self.storage.working_set_bucket
+
+    fn get_next_working_set_number(&self) -> anyhow::Result<usize> {
+        let t = self.get_txn()?;
+        let result: Option<usize> = t
+            .query_row("SELECT COALESCE(MAX(id), 0) FROM working_set", [], |r| {
+                r.get(0)
+            })
+            .optional()
+            .context("Getting highest working set ID")?;
+
+        Ok(result.unwrap_or(0) + 1)
```

Comment:
just for clarity:
```suggestion
            .query_row("SELECT COALESCE(MAX(id), 0) + 1 FROM working_set", [], |r| {
                r.get(0)
            })
            .optional()
            .context("Getting highest working set ID")?;

        Ok(result.unwrap_or(0))
```

Reply:
Yep makes sense - I think that'll might work better also (since if there is no results, the old code would have returned `1` for first ID). Also renamed the terrible `result` variable to something more descriptive